map/pointcloud_divider/include/pointcloud_divider/pointcloud_divider.hpp

```diff
@@ -0,0 +1,140 @@
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef POINTCLOUD_DIVIDER__POINTCLOUD_DIVIDER_HPP_
+#define POINTCLOUD_DIVIDER__POINTCLOUD_DIVIDER_HPP_
+
+#include <yaml-cpp/yaml.h>
+
+#include <string>
+#include <tuple>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#define PCL_NO_PRECOMPILE
+#include "grid_info.hpp"
+#include "pcd_io.hpp"
+
+#include <pcl/filters/voxel_grid.h>
+#include <pcl/io/pcd_io.h>
+
+template <class PointT>
+class PointCloudDivider
+{
+  typedef pcl::PointCloud<PointT> PclCloudType;
+  typedef typename PclCloudType::Ptr PclCloudPtr;
+  typedef std::unordered_map<GridInfo<2>, std::tuple<PclCloudType, int, size_t>> GridMapType;
+  typedef typename GridMapType::iterator GridMapItr;
+  typedef std::multimap<size_t, GridMapItr> GridMapSizeType;
+  typedef typename GridMapSizeType::iterator GridMapSizeItr;
+
+public:
+  PointCloudDivider() {}
+  ~PointCloudDivider() {}
+
+  std::pair<double, double> getGridSize() const
+  {
+    return std::pair<double, double>(grid_size_x_, grid_size_y_);
+  }
+
+  void run(
+    std::vector<std::string> & pcd_names, const std::string & output_dir,
+    const std::string & file_prefix, const std::string & config);
+
+  void run(
+    const PclCloudPtr & cloud, const std::string & output_dir, const std::string & file_prefix,
+    const std::string & config);
+
+  std::string makeFileName(const GridInfo<2> & grid) const;
+
+  void setUseLargeGrid(const bool use_large_grid) { use_large_grid_ = use_large_grid; }
+
+private:
+  PclCloudPtr merged_ptr_;
+  std::string output_dir_, file_prefix_, config_file_;
+
+  std::unordered_set<GridInfo<2>> grid_set_;
+
+  // Params from yaml
+  bool use_large_grid_ = false;
+  bool merge_pcds_ = false;
+  double leaf_size_ = 0.1;
+  double grid_size_x_ = 100;
+  double grid_size_y_ = 100;
+  double g_grid_size_x_ = grid_size_x_ * 10;
+  double g_grid_size_y_ = grid_size_y_ * 10;
+
+  // Maximum number of points per PCD block
+  const size_t max_block_size_ = 500000;
+
+  // Map of points distributed to grids
+  GridMapType grid_to_cloud_;
+
+  // Segments but sorted by size
+  GridMapSizeType seg_by_size_;
+  // Map segments to size iterator
+  std::unordered_map<GridInfo<2>, GridMapSizeItr> seg_to_size_itr_map_;
+  // Only 100 million points are allowed to reside in the main memory at max
+  const size_t max_resident_point_num_ = 100000000;
+  size_t resident_point_num_ = 0;
+  std::string tmp_dir_;
+  CustomPCDReader<PointT> reader_;
+  size_t total_point_num_ = 0;  // Used when mergning all pcds
```

Comment:
please add pcds to [autoware-spell-check-dict](https://github.com/tier4/autoware-spell-check-dict)